### PR TITLE
feat: switch chain to mainnet on wallet re/connect

### PR DESCRIPTION
**Pull request by gomesalexandre**
## Description

This switches the wallet's active chain to Ethereum Mainnet if set to any other chain, on wallet connect from splash screen / wallet reconnect when reloading the app.

## Notice

- [x] Have you followed the guidelines in our [Contributing]('https://github.com/shapeshift/web/CONTRIBUTING.md) guide?
- [x] Have you checked to ensure there aren't other open [Pull Requests](https://github.com/shapeshift/web/pulls) for the same update/change?

## Pull Request Type

- [ ] :bug: Bug fix (Non-breaking Change: Fixes an issue)
- [ ] :hammer_and_wrench: Chore (Non-breaking Change: Doc updates, pkg upgrades, typos, etc..)
- [x] :nail_care: New Feature (Breaking/Non-breaking Change)

## Issue (if applicable)

- closes https://github.com/shapeshift/web/issues/1843

## Risk

None - this is isolated to MetaMask, XDEFI and TallyHo. If the wallet is set to Mainnet, the flow will continue as before. 

### Wallet Connect from splash screen

- Before opening the app, switch your MetaMask/XDEFI wallet to any chain other than mainnet
- Click connect MM/XDEFI
- XDEFI should silently switch to mainnet and continue the flow / MM should bring a network switch request, which will continue with the flow after accepting the request

### App reload

- Go to the app with MM/XDEFI set to mainnet and connect your wallet
- Switch your MetaMask/XDEFI wallet to any chain other than mainnet
- Reload the page
- XDEFI should silently switch to mainnet and continue the flow / MM should bring a network switch request, which will continue with the flow after accepting the request

## Screenshots (if applicable)

image

### CODE REVIEW

File: src/context/WalletProvider/TallyHo/components/Connect.tsx [hunk 1]
```diff
@@ -68,8 +71,10 @@ export const TallyHoConnect = ({ history }: TallyHoSetupProps) => {
       try {
         const deviceId = await wallet.getDeviceID()
 
-        if (provider?.chainId !== '0x1') {
-          throw new Error('walletProvider.tallyHo.errors.network')
+        // Switch to Mainnet if wallet is on any other chain
+        const chainId = await wallet.ethGetChainId?.()
+        if (bnOrZero(chainId).toString() !== CHAIN_REFERENCE.EthereumMainnet) {
+          await wallet.ethSwitchChain?.(bn(CHAIN_REFERENCE.EthereumMainnet).toNumber())
```
Comment (gomesalexandre) [inline src/context/WalletProvider/TallyHo/components/Connect.tsx:77]: Note: This is currently a no-op. I managed to get to a chainId other than 1, and thus to this block, by building Tally from source and enabling all the multichain dev flags:
```
SUPPORT_POLYGON=true
SUPPORT_ARBITRUM=true
SUPPORT_OPTIMISM=true
```
However, `wallet_switchEthereumChain`, while broadcasted, isn't actually picked by Tally yet.
Until this gets released in Tally, this is a dead block, as the condition will always evaluate to false. IMO, it's better to still keep it here for consistency.

File: src/context/WalletProvider/WalletProvider.tsx [hunk 2]
```diff
@@ -394,6 +406,12 @@ export const WalletProvider = ({ children }: { children: React.ReactNode }): JSX
               if (localWalletType === 'tallyho' && window?.ethereum?.isMetaMask) disconnect()
               const localTallyHoWallet = await state.adapters.get(KeyManager.TallyHo)?.pairDevice()
               if (localTallyHoWallet) {
+                const chainId = await localTallyHoWallet.ethGetChainId?.()
+                if (bnOrZero(chainId).toString() !== CHAIN_REFERENCE.EthereumMainnet) {
+                  await localTallyHoWallet.ethSwitchChain?.(
+                    bn(CHAIN_REFERENCE.EthereumMainnet).toNumber(),
+                  )
```
Comment (gomesalexandre) [inline src/context/WalletProvider/WalletProvider.tsx:413]: Note: This is currently a no-op. I managed to get to a chainId other than 1, and thus to this block, by building Tally from source and enabling all the multichain dev flags:
```
SUPPORT_POLYGON=true
SUPPORT_ARBITRUM=true
SUPPORT_OPTIMISM=true
```
However, `wallet_switchEthereumChain`, while broadcasted, isn't actually picked by Tally yet.
Until this gets released in Tally, this is a dead block, as the condition will always evaluate to false. IMO, it's better to still keep it here for consistency.

**Comment by gomesalexandre** [comment]
Moving to draft until new hdwallet version is published.

File: src/context/WalletProvider/MetaMask/components/Connect.tsx [hunk 3]
```diff
@@ -55,7 +58,7 @@ export const MetaMaskConnect = ({ history }: MetaMaskSetupProps) => {
     }
 
     if (state.adapters && state.adapters?.has(KeyManager.MetaMask)) {
-      const wallet = await state.adapters.get(KeyManager.MetaMask)?.pairDevice()
+      const wallet = (await state.adapters.get(KeyManager.MetaMask)?.pairDevice()) as ETHWallet
```
Comment (0xdef1cafe) [inline src/context/WalletProvider/MetaMask/components/Connect.tsx:61]: use metamaskhdwallet over ethwallet

File: src/context/WalletProvider/TallyHo/components/Connect.tsx (hunk 1 above)
Comment (0xApotheosis) [reply src/context/WalletProvider/TallyHo/components/Connect.tsx:77]: Can we add a condensed version of this comment as a version-controlled comment?

File: src/context/WalletProvider/WalletProvider.tsx (hunk 2 above)
Comment (0xApotheosis) [reply src/context/WalletProvider/WalletProvider.tsx:417]: Same here, can we add a comment in the code?

**Review by 0xApotheosis — DISMISSED** [review]
Tested with Metamask and works well 👍

Screen Shot 2022-06-30 at 12 50 29 pm

File: src/context/WalletProvider/WalletProvider.tsx (hunk 2 above)
Comment (gomesalexandre) [reply src/context/WalletProvider/WalletProvider.tsx]: Added in bdb09a0bad87b7e7ffe3ce0a5e0fbdf2bfdf4018 !

**Review by 0xdef1cafe — APPROVED** [review]
🤌

**Review by 0xApotheosis — APPROVED** [review]
🧂

**Pull request merged**